### PR TITLE
Now syncs and modifies in a better way

### PR DESCRIPTION
**Pull request by sveinnthorarins**
I changed the way the plugin syncs the filename and heading resulting in these two limitations being eradicated:

- Doesn't work yet when editor is in preview mode
- When renaming a file that isn't the current file, nothing will happen. The heading will get updated the next time the file is saved

Now the plugin modifies the file by actually reading and writing to the file instead of using the editor to modify. This might slow the syncing process by a bit for large files but I have not tested it. I don't have much experience with Obsidian plugins but I believe this code is pretty straight-forward, nevertheless it would be best if you could test this version extensively.

I tried to thoroughly comment any code that was heavily transformed (mainly the functions findNoteStart, findHeading and replaceLine). 

Let me know what you think 😃

(Oh, also! I used npm to install the packages and upon doing so it changed the yarn.lock file by adding some quotes around package property names, I'm not sure if this is good or not but I included it in the commit anyhow. If you don't want that change, make sure you edit that out! 😛)

### CODE REVIEW

**Comment by sveinnthorarins** [comment]
I'm also fairly confident that with this change it would be easy to introduce the file-open event again. Since now when an event fires the whole file's contents is actually read at that time, instead of working with some file objects from the editor that might be old. That is, if the file has been saved (written to on disk) before a file-open event is fired, no bugs should occur (which I believe was an issue at some point?).

File: main.ts [hunk 1]
```diff
@@ -109,27 +107,22 @@ export default class FilenameHeadingSyncPlugin extends Plugin {
...
-      );
-      this.app.fileManager.renameFile(view.file, newPath);
-    }
+      const sanitizedHeading = this.sanitizeHeading(heading.text);
+      if (
+        sanitizedHeading.length > 0 &&
+        this.sanitizeHeading(file.basename) !== sanitizedHeading
+      ) {
+        const newPath = file.path.replace(file.basename, sanitizedHeading);
+        this.app.vault.rename(file, newPath);
```
Comment (dvcrn) [inline main.ts:123]: I used `renameFile` here because `rename` did not update links in other files. Is that still the case?

File: main.ts [hunk 2]
```diff
@@ -168,69 +154,64 @@ export default class FilenameHeadingSyncPlugin extends Plugin {
...
-      return;
-    }
-
-    if (docstart > 0) {
-      this.insertLine(doc, docstart + 1, `\n# ${sanitizedHeading}`);
-    } else {
-      this.insertLine(doc, docstart, `# ${sanitizedHeading}`);
-    }
-    doc.setCursor(cursor);
+    this.app.vault.read(file).then((data) => {
```
Comment (dvcrn) [inline main.ts:158]: 👍

File: main.ts [hunk 3]
```diff
@@ -168,69 +154,64 @@ export default class FilenameHeadingSyncPlugin extends Plugin {
...
+      const lines = data.split('\n');
+      const start = this.findNoteStart(lines);
+      const heading = this.findHeading(lines, start);
+
+      if (heading !== null) {
+        if (this.sanitizeHeading(heading.text) !== sanitizedHeading) {
+          this.replaceLine(
```
Comment (dvcrn) [inline main.ts:165]: Is replaceLine alone here enough? Does `data` need to get written back manually, or is obsidian handling that already?

File: main.ts (hunk 1 above)
Comment (sveinnthorarins) [reply main.ts:123]: Ah, I very much missed this functionality, that is indeed the case! Will switch this back to the FileManager's renameFile method.

File: main.ts (hunk 3 above)
Comment (sveinnthorarins) [reply main.ts:165]: The `data` needs to be written back manually. I decided to perform the writing in the replaceLine function, to keep our sync function from becoming messy. So, please check out the replaceLine function, it's essentially completely transformed.

Comment (dvcrn) [reply main.ts:165]: Ah okay now I see it! How about `replaceLineInFile`? That makes it more clear what's actually happening. 

Personally also not a fan of the `insertInstead` parameter. Semantically `replaceLine` means *replace*, so having a function that's called `replace` doing an insertion instead is a bit odd. Can we split it into insert/replace?

Comment (sveinnthorarins) [reply main.ts:165]: Yeah, I agree on adding the `InFile`, that makes it much clearer.

I also agree the `insertInstead` parameter is a bit hacky and we could of course split the function in two (insert/replace). However, that way we are duplicating a bit of code, which can also be considered a 'bad practice' and makes it more complicated to modify later on. 

If you are fine with that, then I'll make the change, but can we maybe find a word that serves the purpose of both operations? All I can think of is `editLineInFile` but inserting a line isn't really editing a line, is it? Should we go for a long `addOrEditLineInFile`? 😛

There's also the word 'upsert' used for update/insert in database management, but I'm not sure how I feel about it.

Comment (dvcrn) [reply main.ts:165]: Thought about this a bit and couldn't come up with a better word either. I think it's fine to split and have a bit of duplication, it also makes the functions itself a bit slimmer.

**Comment by dvcrn** [comment]
The code looks fine to me. I'll give this PR a proper try when I'm with a bit more time!

**Comment by dvcrn** [comment]
Found a small issue: 

- Inserts heading over frontmatter when inserting a new heading

**Comment by sveinnthorarins** [comment]
Ah okay, I believe we need to call `.trim()` on the lines before trying to match them to frontmatter's `---`.

I'll make that change immediately.

**Comment by dvcrn** [comment]
Okay this was a mistake on my end. All works fine!

I'll get this merged, thanks for the contribution! ❤️

**Pull request merged**